File: Internship-Android-Markdown/IPC/IPC/aidl/src/main/java/com/xiaoyingbo/idle/MainActivity.java
package com.xiaoyingbo.idle;

import android.app.Application;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.util.Log;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import com.xiaoyingbo.idle.impl.IBookImpl;
import com.xiaoyingbo.idle.impl.IUserImpl;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    IUserInterface iUserInterface ;
    IBookInterface iBookInterface;
    private boolean bound=true;
    private BinderPool binderPool;

    /*线程池*/
    private static ThreadPoolExecutor taskPool = (ThreadPoolExecutor)new ThreadPoolExecutor(16, 32
            , 200L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(1000),
            new ThreadFactory() {
                @Override
                public Thread newThread(Runnable r) {
                    Thread thread = new Thread(r);
                    thread.setName("线程"+String.valueOf(taskPool.getPoolSize())+"号");
                    return thread;
                }
            });

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViewById(R.id.btn_bind_service).setOnClickListener(this);
        findViewById(R.id.btn_setuser).setOnClickListener(this);
        findViewById(R.id.btn_getuser).setOnClickListener(this);
        findViewById(R.id.btn_setbook).setOnClickListener(this);
        findViewById(R.id.btn_getbook).setOnClickListener(this);

        taskPool.execute(new Runnable() {
            @Override
            public void run() {
                binderPool = BinderPool.getInstance(MainActivity.this);
            }
        });
    }
    @Override
    protected void onStop() {
        super.onStop();
        if (bound) {
            bound = false;
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.P)
    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_bind_service:
                bound=true;
                /*BinderPool.getInstance(this).connectBinderPoolService();*/
                taskPool.execute(new Runnable() {
                    @Override
                    public void run() {
                        IBinder iUserBinder = binderPool.queryBinder(BinderPool.BINDER_USER);
                        iUserInterface= IUserImpl.asInterface(iUserBinder);
                        IBinder iBookBinder = binderPool.queryBinder(BinderPool.BINDER_BOOK);
                        iBookInterface= IBookImpl.asInterface(iBookBinder);
                    }
                });
                break;
            case R.id.btn_setuser:
                taskPool.execute(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            iUserInterface.setUser(new User("user",15));
                            Log.e("aidl-demo","MainActivity:"+" "+ Application.getProcessName()+" "+Thread.currentThread().getName());
                        } catch (RemoteException e) {
                            e.printStackTrace();
                        }
                    }
                });
                break;
            case R.id.btn_getuser:
                taskPool.execute(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            User user = iUserInterface.getUser();
                            Log.e("aidl-demo","MainActivity:"+user.toString()+" "+ Application.getProcessName()+" "+Thread.currentThread().getName());
                        } catch (RemoteException e) {
                            e.printStackTrace();
                        }
                    }
                });
                break;
            case R.id.btn_setbook:
                taskPool.execute(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            iBookInterface.setBook(new Book("book"));
                            Log.e("aidl-demo","MainActivity:"+" "+ Application.getProcessName()+" "+Thread.currentThread().getName());
                        } catch (RemoteException e) {
                            e.printStackTrace();
                        }
                    }
                });
                break;
            case R.id.btn_getbook:
                taskPool.execute(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            Book book = iBookInterface.getBook();
                            Log.e("aidl-demo","MainActivity:"+book.toString()+" "+ Application.getProcessName()+" "+ Thread.currentThread().getName());
                        } catch (RemoteException e) {
                            e.printStackTrace();
                        }
                    }
                });
                break;
        }
    }

}

File: Internship-Android-Markdown/IPC/IPC/messenger/src/main/java/com/xiaoyingbo/messenger/MainActivity.kt
package com.xiaoyingbo.messenger

import android.app.Application
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.*
import android.util.Log
import android.view.View
import android.widget.Button
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity


class MainActivity : AppCompatActivity(), View.OnClickListener {
    var remoteMessage:Messenger ?= null/*远程信使*/
    var localMessenger:Messenger ?= null/*本地信使*/
    private var bound:Boolean?= null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViewById<Button>(R.id.btn_bind_service).setOnClickListener(this)
        findViewById<Button>(R.id.btn_send_msg).setOnClickListener(this)
    }

    private var connection:ServiceConnection = object :ServiceConnection {
        override fun onServiceConnected( name:ComponentName,  service:IBinder) {
            //从原始 IBinder 创建一个 Messenger，该 IBinder 之前已使用 getBinder 检索到。
            remoteMessage = Messenger(service);/*从服务端的IBinder中获取到远程信使*/
            bound = true;
        }

        override fun onServiceDisconnected( name:ComponentName) {
            bound = false;
        }
    };


    class ReturnHandler : Handler() {
        @RequiresApi(Build.VERSION_CODES.P)
        override fun handleMessage(msg:Message) {
            when (msg.what) {
                MessengerService.MSG_SET_VALUE->{/*如果handler中接收到的消息时从服务端来的*/
                    Log.e("messenger-demo","Received from service: " + msg.arg1+ ""+Application.getProcessName());
                }
                else->{
                    super.handleMessage(msg)
                }
            }
        }
    }
    override fun onClick( v:View) {
        when (v.id) {
            R.id.btn_bind_service -> {/*开启服务*/
                bindService(Intent (this@MainActivity, MessengerService::class.java),
                    connection,
                    Context.BIND_AUTO_CREATE)/*绑定服务*/
            }
            R.id.btn_send_msg->{/*发送消息给服务端*/
                try {
                    localMessenger = Messenger(ReturnHandler())/*本地信使会从ReturnHandler这个Handler中收到服务端返回的消息*/

                    var msg = Message.obtain(null, MessengerService.MSG_REGISTER_CLIENT)/**从享元池中取出对应客户端发送的消息*/
                    msg.replyTo = localMessenger;/*先发送一条消息注册客户端的信使*/
                    //先发一则消息添加Messenger：msg.replyTo = localMessenger;
                    remoteMessage!!.send(msg);/*远程信使发送给消息*/

                    // Give it some value as an example.
                    msg = Message.obtain(null, MessengerService.MSG_SET_VALUE, this.hashCode(), 0);
                    //传入的arg1值:this.hashCode()
                    remoteMessage!!.send(msg);
                } catch ( e:RemoteException) {
                    e.printStackTrace()
                }
            }

        }
    }
    override fun onStop() {
        super.onStop();
        if (bound == true) {
            unbindService(connection);
            bound = false;
        }
    }
}